pkg/core/mempool/mempool.go

```diff
@@ -126,42 +154,80 @@ func NewMempool(eventBus *eventbus.EventBus, rpcBus *rpcbus.RPCBus, verifier tra
 	return m
 }
 
-// Run spawns the mempool lifecycle routine. The whole mempool cycle is around
-// getting input from the outside world (from input channels) and provide the
-// actual list of the verified txs (onto output channel).
-//
-// All operations are always executed in a single go-routine so no
-// protection-by-mutex needed.
+// Run spawns the mempool lifecycle routines.
 func (m *Mempool) Run(ctx context.Context) {
-	go func() {
-		ticker := time.NewTicker(idleTime)
-		defer ticker.Stop()
-
-		for {
-			select {
-			// rpcbus methods.
-			case r := <-m.sendTxChan:
-				go handleRequest(r, m.processSendMempoolTxRequest, "SendTx")
-			case r := <-m.getMempoolTxsChan:
-				handleRequest(r, m.processGetMempoolTxsRequest, "GetMempoolTxs")
-			case r := <-m.getMempoolTxsBySizeChan:
-				handleRequest(r, m.processGetMempoolTxsBySizeRequest, "GetMempoolTxsBySize")
-			case b := <-m.acceptedBlockChan:
-				m.onBlock(b)
-			case <-ticker.C:
-				m.onIdle()
-			// Mempool terminating.
-			case <-ctx.Done():
-				// m.eventBus.Unsubscribe(topics.Tx, m.txSubscriberID)
-				return
+	// Main Loop
+	go m.Loop(ctx)
+
+	// Loop to drain pendingPropagation and try to propagate transaction
+	go m.propagateLoop(ctx)
+}
+
+// Loop listens for GetMempoolTxs request and topics.AcceptedBlock events.
+func (m *Mempool) Loop(ctx context.Context) {
+	ticker := time.NewTicker(idleTime)
+	defer ticker.Stop()
+
+	for {
+		select {
+		// rpcbus methods.
+		case r := <-m.sendTxChan:
+			// TODO: This handler should be deleted once new wallet is integrated
+			go handleRequest(r, m.processSendMempoolTxRequest, "SendTx")
+		case r := <-m.getMempoolTxsChan:
+			handleRequest(r, m.processGetMempoolTxsRequest, "GetMempoolTxs")
+		case r := <-m.getMempoolTxsBySizeChan:
+			handleRequest(r, m.processGetMempoolTxsBySizeRequest, "GetMempoolTxsBySize")
+		case b := <-m.acceptedBlockChan:
+			m.onBlock(b)
+		case <-ticker.C:
+			m.onIdle()
+		case <-ctx.Done():
+			// Mempool terminating
+			return
+		}
+
+		ticker.Reset(idleTime)
+	}
+}
+
+func (m *Mempool) propagateLoop(ctx context.Context) {
+	for {
+		select {
+		case t := <-m.pendingPropagation:
+			// Ensure we propagate at proper rate
+			if m.limiter != nil {
+				if err := m.limiter.Wait(ctx); err != nil {
+					log.WithError(err).Error("failed to limit rate")
+				}
+			}
```

Comment:
The ordering of Transactions to be propagated should follow an order based on the `Fee` field. This needs to be tackled in a different PR

Reply:
Referenced in #1240